agent/engine/containermetadata/acquire_metadata.go

```diff
@@ -0,0 +1,188 @@
+	// nil pointer exceptions (Could occur if there is some error in the docker api call)
+	if container.Config != nil {
+		imageNameFromConfig = container.Config.Image
+	} else {
+		seelog.Errorf("Failed to acquire container metadata: container has no configuration")
+		return DockerContainerMD{}
+	}
+
+	// Get Port bindings from docker configurations
+	var ports []api.PortBinding
+	var err error
+	// A real world container should never lack a host config but we check just in case
+	if container.HostConfig != nil {
+		ports, err = api.PortBindingFromDockerPortBinding(container.HostConfig.PortBindings)
+		if err != nil {
+			seelog.Errorf("Failed to acquire port binding metadata: %v", err)
+		}
+	} else {
+		seelog.Errorf("Failed to acquire container metadata: container has no host configuration")
+		return DockerContainerMD{}
+	}
+
+	networkMD := acquireNetworkMetadata(container.NetworkSettings)
+
+	dockerContainerMD := DockerContainerMD{
+		containerID:   container.ID,
+		containerName: container.Name,
+		imageID:       container.Image,
+		imageName:     imageNameFromConfig,
+		ports:         ports,
+		networkInfo:   networkMD,
+	}
+	return dockerContainerMD
+}
+
+// acquireTaskMetadata parses metadata in the AWS configuration and task
+// and packages this data for JSON marshaling
+// Since we accept incomplete metadata fields, we should not return
+// errors here and handle them at this stage.
+func acquireTaskMetadata(client dockerDummyClient, cfg *config.Config, task *api.Task) TaskMetadata {
+	// Get docker version from client. May block metadata file updates so the file changes
+	// should be made in a goroutine as this does a docker client call
+	version, err := client.Version()
+	if err != nil {
+		version = ""
+		seelog.Errorf("Failed to get docker version number: %v", err)
+	}
+
+	clusterArnFromConfig := ""
+	if cfg != nil {
+		clusterArnFromConfig = cfg.Cluster
+	} else {
+		// This error should not happen in real world use cases (Or occurs somewhere else long
+		// before this should matter)
+		seelog.Errorf("Failed to get cluster Arn: invalid configuration")
+	}
+
+	taskArnFromConfig := ""
+	if task != nil {
+		taskArnFromConfig = task.Arn
+	} else {
+		// This error should not happen in real world use cases (Or occurs somewhere else long
+		// before this should matter)
+		seelog.Errorf("Failed to get task Arn: invalid task")
+	}
+
+	return TaskMetadata{
+		version:    version,
+		clusterArn: clusterArnFromConfig,
+		taskArn:    taskArnFromConfig,
+	}
+}
+
+// acquireMetadata gathers metadata from a docker container, and task
+// configuration and data then packages it for JSON Marshaling
+// Since we accept incomplete metadata fields, we should not return
+// errors here and handle them at this stage.
+func (manager *metadataManager) acquireMetadata(createTime string, updateTime string, container *docker.Container, task *api.Task) Metadata {
+	taskMD := acquireTaskMetadata(manager.client, manager.cfg, task)
```

Comment:
This looks like a mix of creating a Metadata object through a contract and initializing some of its values. If this is for the purpose of serializing is there a way to treat this as a DTO and add serialization as a different method?